File: ios/KidSpeak/KSPurchaseManager.swift
//
//  KSPurchaseManager.swift
//  KidSpeak
//
//  Native IAP Implementation using StoreKit 2
//

import Foundation
import StoreKit

@objc(KSPurchaseManager)
class KSPurchaseManager: RCTEventEmitter {
  
  // Product IDs
  private let productIDs = [
    "com.kidspeak.mobile.weeklytrial1",
    "com.kidspeak.mobile.weekly1",
    "com.kidspeak.mobile.monthly1"
  ]
  
  // Available products
  private var products: [Product] = []
  
  // Transaction listener task
  private var transactionListener: Task<Void, Error>? = nil
  
  // MARK: - Initialization
  
  override init() {
    super.init()
    startTransactionListener()
  }
  
  deinit {
    transactionListener?.cancel()
  }
  
  // MARK: - Transaction Listener
  
  private func startTransactionListener() {
    transactionListener = Task {
      for await result in Transaction.updates {
        if case .verified(let transaction) = result {
          await transaction.finish()
        }
      }
    }
  }
  
  // MARK: - React Native Bridge Methods
  
  @objc
  func initialize(_ resolve: @escaping RCTPromiseResolveBlock,
                  rejecter reject: @escaping RCTPromiseRejectBlock) -> Void {
    Task { @MainActor in
      do {
        // Fetch products from App Store
        self.products = try await Product.products(for: self.productIDs)
        
        let productsData = self.products.map { product in
          return [
            "id": product.id,
            "displayName": product.displayName,
            "displayPrice": product.displayPrice,
            "price": NSNumber(value: product.price),
            "description": product.description
          ] as [String : Any]
        }
        
        resolve(productsData)
      } catch {
        reject("INIT_ERROR", "Failed to initialize IAP: \(error.localizedDescription)", error)
      }
    }
  }
  
  @objc
  func purchaseProduct(_ productId: String,
                       resolver resolve: @escaping RCTPromiseResolveBlock,
                       rejecter reject: @escaping RCTPromiseRejectBlock) -> Void {
    Task { @MainActor in
      guard let product = self.products.first(where: { $0.id == productId }) else {
        reject("PRODUCT_NOT_FOUND", "Product not found: \(productId)", nil)
        return
      }
      
      do {
        let result = try await product.purchase()
        
        switch result {
        case .success(let verification):
          switch verification {
          case .verified(let transaction):
            await transaction.finish()
            
            let expirationTimestamp: NSNumber
            if let expirationDate = transaction.expirationDate {
              expirationTimestamp = NSNumber(value: expirationDate.timeIntervalSince1970)
            } else {
              expirationTimestamp = NSNumber(value: 0)
            }
            
            let purchaseData: [String: Any] = [
              "productId": transaction.productID,
              "transactionId": String(transaction.id),
              "purchaseDate": NSNumber(value: transaction.purchaseDate.timeIntervalSince1970),
              "expirationDate": expirationTimestamp
            ]
            
            resolve(purchaseData)
            
          case .unverified(let transaction, let error):
            await transaction.finish()
            reject("VERIFICATION_FAILED", "Transaction verification failed: \(error.localizedDescription)", nil)
          }
          
        case .userCancelled:
          reject("USER_CANCELLED", "User cancelled the purchase", nil)
          
        case .pending:
          reject("PENDING", "Purchase is pending approval", nil)
          
        @unknown default:
          reject("UNKNOWN_ERROR", "Unknown purchase result", nil)
        }
      } catch {
        reject("PURCHASE_ERROR", "Purchase failed: \(error.localizedDescription)", error)
      }
    }
  }
  
  @objc
  func restorePurchases(_ resolve: @escaping RCTPromiseResolveBlock,
                        rejecter reject: @escaping RCTPromiseRejectBlock) -> Void {
    Task { @MainActor in
      do {
        try await AppStore.sync()
        
        var restoredPurchases: [[String: Any]] = []
        
        for await result in Transaction.currentEntitlements {
          if case .verified(let transaction) = result {
            let expirationDate = transaction.expirationDate ?? Date()
            if expirationDate > Date() {
              let expirationTimestamp: NSNumber
              if let expDate = transaction.expirationDate {
                expirationTimestamp = NSNumber(value: expDate.timeIntervalSince1970)
              } else {
                expirationTimestamp = NSNumber(value: 0)
              }
              
              let purchaseData: [String: Any] = [
                "productId": transaction.productID,
                "transactionId": String(transaction.id),
                "purchaseDate": NSNumber(value: transaction.purchaseDate.timeIntervalSince1970),
                "expirationDate": expirationTimestamp
              ]
              
              restoredPurchases.append(purchaseData)
            }
          }
        }
        
        let resultData: [String: Any] = [
          "success": true,
          "purchases": restoredPurchases,
          "count": restoredPurchases.count
        ]
        resolve(resultData)
      } catch {
        reject("RESTORE_ERROR", "Failed to restore purchases: \(error.localizedDescription)", error)
      }
    }
  }
  
  @objc
  func checkActiveSubscription(_ resolve: @escaping RCTPromiseResolveBlock,
                                rejecter reject: @escaping RCTPromiseRejectBlock) -> Void {
    Task { @MainActor in
      var hasActiveSubscription = false
      var activeProductId: String? = nil
      
      for await result in Transaction.currentEntitlements {
        if case .verified(let transaction) = result {
          let expirationDate = transaction.expirationDate ?? Date()
          if expirationDate > Date() {
            hasActiveSubscription = true
            activeProductId = transaction.productID
            break
          }
        }
      }
      
      let resultData: [String: Any] = [
        "hasActive": hasActiveSubscription,
        "productId": activeProductId as Any
      ]
      resolve(resultData)
    }
  }
  
  // Required for RCTEventEmitter
  override func supportedEvents() -> [String]! {
    return []
  }
  
  override static func requiresMainQueueSetup() -> Bool {
    return false
  }
}

File: android/app/src/main/java/com/kidspeak/KSPurchaseManager.kt
package com.kidspeak

import android.app.Activity
import com.android.billingclient.api.*
import com.facebook.react.bridge.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class KSPurchaseManager(reactContext: ReactApplicationContext) : 
    ReactContextBaseJavaModule(reactContext), PurchasesUpdatedListener {

    private var billingClient: BillingClient? = null
    private val productIds = listOf(
        "com.kidspeak.mobile.weeklytrial1",
        "com.kidspeak.mobile.weekly1",
        "com.kidspeak.mobile.monthly1"
    )
    private var products: List<ProductDetails> = emptyList()
    
    override fun getName(): String = "KSPurchaseManager"

    init {
        initializeBillingClient()
    }

    private fun initializeBillingClient() {
        billingClient = BillingClient.newBuilder(reactApplicationContext)
            .setListener(this)
            .enablePendingPurchases()
            .build()
    }

    @ReactMethod
    fun initialize(promise: Promise) {
        billingClient?.startConnection(object : BillingClientStateListener {
            override fun onBillingSetupFinished(billingResult: BillingResult) {
                if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                    // Query products
                    queryProducts(promise)
                } else {
                    promise.reject(
                        "INIT_ERROR",
                        "Billing client setup failed: ${billingResult.debugMessage}"
                    )
                }
            }

            override fun onBillingServiceDisconnected() {
                // Try to reconnect
                initializeBillingClient()
            }
        })
    }

    private fun queryProducts(promise: Promise) {
        val productList = productIds.map { productId ->
            QueryProductDetailsParams.Product.newBuilder()
                .setProductId(productId)
                .setProductType(BillingClient.ProductType.SUBS)
                .build()
        }

        val params = QueryProductDetailsParams.newBuilder()
            .setProductList(productList)
            .build()

        billingClient?.queryProductDetailsAsync(params) { billingResult, productDetailsList ->
            if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                products = productDetailsList
                
                val productsArray = Arguments.createArray()
                productDetailsList.forEach { product ->
                    val productMap = Arguments.createMap().apply {
                        putString("id", product.productId)
                        putString("displayName", product.name)
                        putString("description", product.description)
                        
                        product.subscriptionOfferDetails?.firstOrNull()?.let { offer ->
                            offer.pricingPhases.pricingPhaseList.firstOrNull()?.let { phase ->
                                putString("displayPrice", phase.formattedPrice)
                                putDouble("price", phase.priceAmountMicros / 1000000.0)
                            }
                        }
                    }
                    productsArray.pushMap(productMap)
                }
                
                promise.resolve(productsArray)
            } else {
                promise.reject(
                    "QUERY_ERROR",
                    "Failed to query products: ${billingResult.debugMessage}"
                )
            }
        }
    }

    @ReactMethod
    fun purchaseProduct(productId: String, promise: Promise) {
        val activity: Activity? = reactApplicationContext.currentActivity
        if (activity == null) {
            promise.reject("NO_ACTIVITY", "Activity not available")
            return
        }

        val product = products.firstOrNull { it.productId == productId }
        if (product == null) {
            promise.reject("PRODUCT_NOT_FOUND", "Product not found: $productId")
            return
        }

        val offerToken = product.subscriptionOfferDetails?.firstOrNull()?.offerToken
        if (offerToken == null) {
            promise.reject("NO_OFFER", "No offer available for product")
            return
        }

        val productDetailsParams = BillingFlowParams.ProductDetailsParams.newBuilder()
            .setProductDetails(product)
            .setOfferToken(offerToken)
            .build()

        val billingFlowParams = BillingFlowParams.newBuilder()
            .setProductDetailsParamsList(listOf(productDetailsParams))
            .build()

        // Store promise for later use in onPurchasesUpdated
        pendingPromise = promise

        val result = billingClient?.launchBillingFlow(activity, billingFlowParams)
        if (result?.responseCode != BillingClient.BillingResponseCode.OK) {
            pendingPromise = null
            promise.reject(
                "LAUNCH_ERROR",
                "Failed to launch billing flow: ${result?.debugMessage}"
            )
        }
    }

    private var pendingPromise: Promise? = null

    override fun onPurchasesUpdated(
        billingResult: BillingResult,
        purchases: MutableList<Purchase>?
    ) {
        val promise = pendingPromise ?: return
        pendingPromise = null

        when (billingResult.responseCode) {
            BillingClient.BillingResponseCode.OK -> {
                purchases?.forEach { purchase ->
                    if (purchase.purchaseState == Purchase.PurchaseState.PURCHASED) {
                        // Acknowledge purchase
                        acknowledgePurchase(purchase, promise)
                    }
                }
            }
            BillingClient.BillingResponseCode.USER_CANCELED -> {
                promise.reject("USER_CANCELLED", "User cancelled the purchase")
            }
            else -> {
                promise.reject(
                    "PURCHASE_ERROR",
                    "Purchase failed: ${billingResult.debugMessage}"
                )
            }
        }
    }

    private fun acknowledgePurchase(purchase: Purchase, promise: Promise) {
        if (purchase.isAcknowledged) {
            resolvePurchase(purchase, promise)
            return
        }

        val params = AcknowledgePurchaseParams.newBuilder()
            .setPurchaseToken(purchase.purchaseToken)
            .build()

        billingClient?.acknowledgePurchase(params) { billingResult ->
            if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                resolvePurchase(purchase, promise)
            } else {
                promise.reject(
                    "ACKNOWLEDGE_ERROR",
                    "Failed to acknowledge purchase: ${billingResult.debugMessage}"
                )
            }
        }
    }

    private fun resolvePurchase(purchase: Purchase, promise: Promise) {
        val purchaseMap = Arguments.createMap().apply {
            putString("productId", purchase.products.firstOrNull() ?: "")
            putString("transactionId", purchase.orderId ?: "")
            putDouble("purchaseDate", purchase.purchaseTime.toDouble())
            putString("purchaseToken", purchase.purchaseToken)
        }
        promise.resolve(purchaseMap)
    }

    @ReactMethod
    fun restorePurchases(promise: Promise) {
        val params = QueryPurchasesParams.newBuilder()
            .setProductType(BillingClient.ProductType.SUBS)
            .build()

        billingClient?.queryPurchasesAsync(params) { billingResult, purchases ->
            if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                val purchasesArray = Arguments.createArray()
                
                purchases.forEach { purchase ->
                    if (purchase.purchaseState == Purchase.PurchaseState.PURCHASED) {
                        val purchaseMap = Arguments.createMap().apply {
                            putString("productId", purchase.products.firstOrNull() ?: "")
                            putString("transactionId", purchase.orderId ?: "")
                            putDouble("purchaseDate", purchase.purchaseTime.toDouble())
                            putString("purchaseToken", purchase.purchaseToken)
                        }
                        purchasesArray.pushMap(purchaseMap)
                    }
                }

                val resultMap = Arguments.createMap().apply {
                    putBoolean("success", true)
                    putArray("purchases", purchasesArray)
                    putInt("count", purchases.size)
                }
                promise.resolve(resultMap)
            } else {
                promise.reject(
                    "RESTORE_ERROR",
                    "Failed to restore purchases: ${billingResult.debugMessage}"
                )
            }
        }
    }

    @ReactMethod
    fun checkActiveSubscription(promise: Promise) {
        val params = QueryPurchasesParams.newBuilder()
            .setProductType(BillingClient.ProductType.SUBS)
            .build()

        billingClient?.queryPurchasesAsync(params) { billingResult, purchases ->
            if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
                val activePurchase = purchases.firstOrNull { 
                    it.purchaseState == Purchase.PurchaseState.PURCHASED 
                }
                
                val resultMap = Arguments.createMap().apply {
                    putBoolean("hasActive", activePurchase != null)
                    if (activePurchase != null) {
                        putString("productId", activePurchase.products.firstOrNull() ?: "")
                    } else {
                        putNull("productId")
                    }
                }
                promise.resolve(resultMap)
            } else {
                promise.reject(
                    "CHECK_ERROR",
                    "Failed to check subscription: ${billingResult.debugMessage}"
                )
            }
        }
    }
}

